File: pset6/credit/credit.py
from cs50 import get_int

# get the credit card number from the user
while True:
    credit_number = get_int("Number: ")
    # if the number is less than zero force to get
    if credit_number >= 0:
        break

# save card number to no in order to avoid losing number
no = credit_number

# declare global variable check_sum
check_sum = 0

# sum every odd number
# start w the last digit, then devide to 100 to get next odd no
# do it until no < 0
while no > 0:
    last_digit = int(no % 10)
    check_sum = check_sum + last_digit
    no = no / 100

# assign number / 10 to no in order to get the even numbers
no = credit_number / 10

# sum every even number
# start last (which is now even no) // devide 100 ...
while no > 0:
    last_digit = int(no % 10)
    check_sum = check_sum + (last_digit * 2)
    check_sum = check_sum + int((last_digit * 2) / 10)
    no = no / 100

# get the last digit of check_sum
check_sum = check_sum % 10

# calculate last numbers of the card
amex = int(credit_number / 10**13)
mastercard = int(credit_number / 10**14)
visa1 = int(credit_number / 10**12)
visa2 = int(credit_number / 10**15)

# print each card if associated

if amex/10 > 0 and amex/10 < 10 and check_sum == 0 and (amex == 34 or amex == 37):
    print("AMEX")

elif mastercard/10 > 0 and mastercard/10 < 10 and check_sum == 0 and (mastercard == 51 or mastercard == 52 or mastercard == 53 or mastercard == 54 or mastercard == 55):
    print("MASTERCARD")

elif visa2/10 > 0 and visa2/10 < 10 and check_sum == 0 and visa2 == 4:
    print("VISA")

elif visa1/10 > 0 and visa1/10 < 10 and check_sum == 0 and visa1 == 4:
    print("VISA")

else:
    print("INVALID")

File: pset1/credit/credit.c
#include <stdio.h>
#include <cs50.h>

int main(void)

{
    
    long number;                                                                                //prompt to get a credit card number
    do
    {
        number = get_long("Number: ");
    }
    while (number <= 0);                                                                          //force to get a number > 0 if not
    
    int checksum = 0;                                                                           // checksum starts w/ zero
    long no = number;                                                                           // number = no; to avoid miscalculation; otherwise it'll fail to calculate even numbers
    
    while (no > 0)                                                                            // get int last digit, and 3rd .. odds
    {
        int lastdigit = no % 10;
        checksum = checksum + lastdigit;
        no = no / 100;
    }
    
    no = number / 10;                                                                    //to get even digits; starts w/ last 2nd digit, and 4th....
    
    while (no > 0)
    {
        int lastdigit = no % 10;
        checksum = checksum + (lastdigit * 2);
        checksum = checksum + ((lastdigit * 2) / 10);                                          // if 2x even number has two digits, to get first digit
        no = no / 100;
    }
    
    checksum = checksum % 10;                                                           // to get the last digit of the checksum, then we'll check if it's == 0
                                                                               
    // Declare ints for AmEx; MasterCard;Visa
    
    int amex = number / 1e+13;                                                                                  
    int mcard = number / 1e+14;
    int visa1 = number / 1e+12;                                                           //it's for 13 digits visa number
    int visa2 = number / 1e+15;                                                           //it's for 16 digits visa number
    
    // int 4 american express' first 2 digits --> start w/ 34||37; 15digits; checksum = 0
    
    if ((amex/10 > 0) && (amex/10 < 10) && (checksum == 0) && (amex == 34 || amex == 37))  
    {
        printf("AMEX\n");
    }
    // int 4 mastercard's first 2 digits --> start w/ 51, 52, 53, 54, 55 
    
    else if ((mcard / 10 > 0) && (mcard / 10 < 10) && (checksum == 0) && (mcard == 51 || mcard == 52 || mcard == 53 || mcard == 54 || mcard == 55))
    {
        printf("MASTERCARD\n");
    }
    
    // visa 13 digits or 16 digits; start w/ 4
    else if (((visa2 > 0) && (visa2 < 10) && (checksum == 0) && (visa2 == 4)) || ((visa1 > 0) && (visa1 < 10) && (checksum == 0) && (visa1 == 4)))
    {
        printf("VISA\n");
    }
    
    else
    {
        printf("INVALID\n");
    }
}